Make DummyRange work this types other than uint

burner opened the pull request:
* fixating the current DummyRange behaviour in a unittest.
* structured the tests a little
* DummyRange with custom types
* DummyRange line coverage from < 10% to 100%
* DummyRange and AllDummyRanges bahaviour have unchanged interface and semantics

andralex commented on std/internal/test/dummyrange.d:
```diff
@@ -28,23 +28,40 @@ enum ReturnBy
 // numbers 1..10.
-struct DummyRange(ReturnBy _r, Length _l, RangeType _rt)
+struct DummyRange(ReturnBy _r, Length _l, RangeType _rt, T = uint[])
+        if(isArray!T)
```
space after if, sigh

andralex commented on std/internal/test/dummyrange.d:
```diff
@@ -28,23 +28,40 @@ enum ReturnBy
 
-    uint[] arr = [1U, 2U, 3U, 4U, 5U, 6U, 7U, 8U, 9U, 10U];
+    //static if (isArray!T && is(Unqual!(ElementType!T) == uint))
+    static if (is(T == uint[]))
```
... like here. Really, inconsistent formatting within the same PR cannot go through.

burner replied:
all if's and for(each)'s not are followed by an space

9il commented on std/internal/test/dummyrange.d:
```diff
@@ -28,23 +28,41 @@ enum ReturnBy
     {
         // Workaround for DMD bug 4378
-        arr = [1U, 2U, 3U, 4U, 5U, 6U, 7U, 8U, 9U, 10U];
+        static if (is(T == uint[]))
```
How this would work for int?

9il replied:
Do `int` and `uint` dummy ranges have different behaviour?

burner replied:
The behavior is not different, but the original DummyRange was an uint[] only (1 to 11) range, and I wanted to preserve that bahavior so the DummyRange api would not break. But reinit broke every DummyRange(T) where a uint[] could not implicitliy be assigned to a T[]

quickfur commented on std/internal/test/dummyrange.d:
```diff
@@ -271,4 +307,251 @@ unittest
+
+    import std.meta : AliasSeq;
+
+    foreach (S; AliasSeq!(uint,double,TestFoo))
```
More style nitpicks: please put a space after `,`, here and also elsewhere in this PR. It makes the result easier to read.

9il commented:
> Wait, so this is adding tests for an internal unit test helper that's already being testing from other unit tests? Or am I missing something?

@burner added new functionality, it should be tested. In addition, it is an example for Phobos contributors.
D Style :)

burner commented:
@klickverbot this module was used to test other modules, it's own behavior was not tested.
I added the possible to use different types as uint with it. Before I added to possibility to use different types but uint, I added tests to make sure I did not change some behavior unused in phobos.

burner commented:
@9il I do not follow

quickfur commented:
ping
Where are we going with this?  Since this is only updating internal test ranges, I'm inclined to just merge it.  Any objections?  @andralex @klickverbot @9il

9il commented:
LGTM

burner commented:
as said earlier, the tests where added to fixate the api and behavior of the original DummyRange implementation, before I added the new functionality to DummyRange.

So it is not just tests for tests,

9il commented:
Auto-merge toggled on

burner commented:
I would argue, that long term having DummyRange!T will yield better tests for other functions and it will fixing possible bugs in DummyRange easier.

I will write a test with iota to test Cmp(T) and TestFoo ...

burner commented:
@9il thanks
